### ISSUE TITLE
Problem when testing your solution

### ISSUE THREAD

**Issue by MVilstrup** [tca19/near-lossless-binarization#3]
Hey Julien Tissier, Christophe Gravier and Amaury Habrard

Your paper looks really interesting and I wanted to test the binary vectors. However, when I run the command `./binarize -input ../crawl-300d-2M.vec` on the [facebook vectors](https://fasttext.cc/docs/en/english-vectors.html ) I get the result on a Ubuntu 16.04 server with openblas-dev installed:

```
1999995 256

, 12854058564283852332 14495264440446430917 2338217189150912797 3900823073646626759

the 2543537490569830019 7091594523546908885 12132390578482067676 17054061891972847710

. 49085064044372029 13430128919309794722 3676268525978997720 1129446822182950747

and 16350715358351243055 16393012712704504925 969627095091772476 2240764738555920805

to 9099634481703745995 17642094685701064884 6452519320616550255 16120898980517001725

of 236520170687141399 4767709100272182524 12135506450154551421 17634123495325585581

a 11194259637020728784 4228149066443023844 7425191957517212067 12735785767725396608

....
```

Futhermore, when I then run the command: ` ./similarity_binary binary_vectors.vec ` on the result, this in turn results in the following error: 

```
create_vocab(): 0.012973s
*** stack smashing detected ***: ./similarity_binary terminated
Aborted (core dumped)
```

I've tried to debug the problem, but have not been able to find the source. Am I doing something wrong?

**Comment by kaierlong** [comment]
@MVilstrup 
thers was an error in your command.
`./similarity_binary` was not followed by a word, but the embeddings file path you have  been binarized.
like,
`./similarity_binary binary_vectors.vec`

**Comment by tca19** [comment]
Hello @MVilstrup,

Thank you for your interest in my work. You are doing it correctly. For the first part of your issue, the command `./binarize -input ../crawl-300d-2M.vec` is correct, and the first lines of `binary_vectors.vec` that you have pasted are actually the binary vectors of some words.
The first line:

`1999995 256` 

indicates the number of word vectors in the file, and the number of bits for each binary vector, in this case 256 bits per vector. The other lines represent the binary vectors, but in a special format. Instead of writing explicitly the binary vector as strings of zeros (0) and ones (1), which would require 256 characters, I am saving the binary vector as a group of integers. Each integer has to be considered as an `unsigned long int`, which is 64 bits. Since there are 4 integers per line, it represents 4 * 64 bits = 256 bits.
For example, the binary vector of the word '_the_' starts with "00100011..." because the binary representation of 2543537490569830019 is 0010001101001100011101011101111100010110010011011001111010000011. We can create the entire binary vector of the word '_the_' by concatenating the binary representation of the integers on the rest of the line.
I save (and load) the binary vectors this way because:
* it needs less characters to write them as integers than to write explicitly a string with 0 or 1 (around 80 characters for the first case, 256 for the second), so `binary_vectors.vec` is smaller.
* the binary operations (like XOR or AND) can only be done on integers, which would require an additional conversion step if it was saved as a binary string.

I will add these explanations into the README for clarification.

For the second part of the issue, you are also correct. The command is indeed `./similarity_binary binary_vectors.vec` (forget the answer of @kaierlong , what is said here is not correct).
I developed and tested my code on Debian sid and this error is not present. Here is the output of this command on my system:
```
create_vocab(): 0.012759s
load_vectors(): 1.119178s
Filename     | Spearman | OOV
==============================
MEN.txt      |    0.768 |   0%
WS353.txt    |    0.694 |   0%
SimLex.txt   |    0.447 |   0%
RW.txt       |    0.511 |   1%
SimVerb.txt  |    0.358 |   0%
evaluate(): 0.005776s
```
As you can see, this command is only needed to evaluate the semantic similarity performance of the binary vectors, not to generate them.
I ran the same command on an Ubuntu 16.04 system and was able to replicate your error. A quick way to overcome this problem is to add the flag `-fno-stack-protector` to `CFLAGS` in the *makefile*, and then run:

 `make clean && make`

The command should now run without errors. The problem seems to be a memory overflow but I don't know where it is so I will leave this issue open until I find and fix it.

**Comment by tca19** [comment]
@MVilstrup , @kaierlong 

The probleme was caused by a segmentation fault. When loading the binary vectors from a file into the memory to compute their performance on semantic similarity datasets, the words are read with `fscanf` and a char buffer of size 64. But some word embeddings (like the ones from Facebook) contain some garbage words, like the word '_SearchNamespacesPageDiscussionViewsViewEditHistoryModelNewsBlogCommunitiesDirectoryStart_' which has a length 88. Therefore, when reading it, it was causing a buffer overflow because the buffer was not big enough to hold it, and making the program crash. I fixed the probleme by increasing the buffer size to 256.

I tested the code on:
* Ubuntu 16.04 with GCC 5.4.0
* Debian sid with GCC 9.2.1

and everything works fine. I am closing this issue.
